eCommerce-Application-Sprint#3

miroslav-zarenkov opened the pull request:
1. Task: [link](https://github.com/rolling-scopes-school/tasks/blob/master/tasks/eCommerce-Application/Sprints/Sprint%233.md)
2. Screenshot:
![screencapture-mir-org-github-io-processors-2023-09-05-15_27_16](https://github.com/mir-org/eCommerce-Application/assets/60261308/dc4ca5e5-4155-45ff-822b-a1b18ec0d3cc)
3. Deploy: [link](https://mir-org.github.io/eCommerce-Application/)
4. Done 05.08.2023 / deadline 05.08.2023
5. Score: 415 / 430
## Кросс-чек (370 points)
### Catalog Page Implementation (150 points)
#### Display Product List (45 points)
- [x] (+25) Use the commercetools API to fetch a list of products with essential details, such as name, image, and description.
- [x] (+20) Display prices with and without discount for discounted products, ensuring that the discounted price is visually distinct and clearly indicates that it is the current price.
#### Product Filtering, Sorting, and Searching (65 points)
- [x] (+30) Utilize the commercetools API to offer robust filtering options for users to refine the product list based on attributes such as price range, brand, color, size, or other relevant characteristics.
- [x] (+15) Enable users to sort the product list by various properties, such as price and name.
- [x] (+20) Implement an efficient and user-friendly search feature that allows users to quickly find and display relevant products based on their search query using the chosen API.
#### Interactive Product Cards (15 points)
- [x] (+10) Design product cards that change their appearance when the user hovers over them, enhancing the browsing experience.
- [x] (+5) Upon clicking a product card, redirect the user to a detailed product information page for the selected product.
#### Category Navigation (25 points)
- [x] (+25) Implement easy-to-use and clear navigation options for users to explore and switch between different product categories or subcategories using the commercetools API.
### Detailed Product Page Implementation (100 points)
#### Display Product Information (65 points)
- [x] (+25) Use the commercetools API to fetch and display the product name, description, and images on the Detailed Product page.
- [x] (+25) Implement an image slider for product images fetched from the chosen API, allowing users to view multiple images of the product.
- [x] (+15) Display the product price fetched from the chosen API, and if the product is on sale, display both the original and discounted prices.
#### Enlarged Image Modal with Slider (35 points)
- [x] (+20) Allow users to click on the product image to open an enlarged version of the image in a modal window.
- [x] (+15) Enable users to navigate through all product images from the commercetools API using a slider inside the modal window.
### User Profile Page Implementation (70 points)
#### Display User Profile Information (30 points)
- [x] (+10) Present the user's personal information, including first name, last name, date of birth, and a list of their addresses in the User Profile page.
- [x] (+20) Provide a user-friendly interface for users to switch to an edit mode, where they can update their personal details, email, and addresses.
#### Edit User Profile Information (40 points)
- [x] (+10) In the edit mode, allow users to update their personal information, including first name, last name, and date of birth and email.
- [x] (+15) Enable users to change their password.
- [ ] (+15) Allow users to manage their addresses, including adding new addresses, deleting existing ones, and updating address details.
### Routing Implementation (40 points)
- [x] (+25) Implement routing for navigation between Catalog page, Product detail page.
- [x] (+15) Implement routing for navigation to User Profile page.
### Evaluation Criteria for Header (10 points)
- [x] (+5) Consistent header layout with necessary information like branding and user navigation across all the pages.
- [x] (+5) User Profile link or button in the header, enabling navigation to User Profile page.
## Проверка ментором (60 points)
- [x] (+10) Code Review Participation.
- [x] (+10) No Merges Without Approvals.
- [x] (+10) Pull Request Descriptions.
- [x] (+10) Task Board Management.
- [x] (+5) Active Participation in Development Discussions.
- [x] (+15) Unit Test Coverage.

toronto-tokyo committed:
3.2.0.add catalog page (#26)

## PR Type

What kind of change does this PR introduce?

```
- [x] Feature
- [ ] Bugfix
- [ ] Refactoring
- [ ] Documentation changes
- [ ] Tests
- [ ] Other
```

## What's new?

Add catalog page

miroslav-zarenkov committed:
3.2.catalog page (#27)

## PR Type

What kind of change does this PR introduce?

```
- [x] Feature
- [ ] Bugfix
- [x] Refactoring
- [ ] Documentation changes
- [ ] Tests
- [ ] Other
```

## What's new?

Add catalog page, add product api, refactor some files.

toronto-tokyo committed:
3.2.2.product filter sort search (#28)

## PR Type

What kind of change does this PR introduce?

```
- [x] Feature
- [ ] Bugfix
- [ ] Refactoring
- [ ] Documentation changes
- [ ] Tests
- [ ] Other
```

## What's new?

Add catalog's sidebar for products searching, sorting, filtering

GinezisNWD committed:
3.1.fix issues (#29)

## PR Type

What kind of change does this PR introduce?

```
- [ ] Feature
- [x] Bugfix
- [ ] Refactoring
- [ ] Documentation changes
- [ ] Tests
- [ ] Other
```

## What's new?
Из карточки ошибок не пофиксил только баг после редиректов, сделаю потом
когда будет больше контента, если вообще сделаю. Переделал хедер,
добавил Observer, на будущее добавил роутеру возможность работать с id,
отрефакторил странички регистрации и логина под работу с Observer.

miroslav-zarenkov committed:
3.2.3.product-cards-hotfix (#30)

## PR Type

What kind of change does this PR introduce?

```
- [x] Feature
- [x] Bugfix
- [ ] Refactoring
- [ ] Documentation changes
- [ ] Tests
- [ ] Other
```

## What's new?

Add id data-attribute to catalog cards.

miroslav-zarenkov committed:
3.2.5.filters on page data (#31)

## PR Type

What kind of change does this PR introduce?

```
- [x] Feature
- [ ] Bugfix
- [x] Refactoring
- [ ] Documentation changes
- [ ] Tests
- [ ] Other
```

## What's new?

Remade pagination and how data loaded on page.

toronto-tokyo committed:
3.2.4.catalog categorization (#33)

## PR Type

What kind of change does this PR introduce?

```
- [x] Feature
- [ ] Bugfix
- [ ] Refactoring
- [ ] Documentation changes
- [ ] Tests
- [ ] Other
```

## What's new?

Add processors and graphic cards categories.

pyropetrick commented on src/api/product-api/product-api-types.ts:
```diff
@@ -0,0 +1,167 @@
+  key: string;
+  lastModifiedAt: string;
+  masterVariant: MasterVariant;
+  metaDescription: {
```
maybe use 1 option type for Localization  - type Locale = Record<LocaleKeys, string> ?

pyropetrick commented on src/api/product-api/product-api-types.ts:
```diff
@@ -0,0 +1,167 @@
+};
+
+type Language = {
+  [language: string]: string;
```
Record<string, string>

pyropetrick commented on src/api/product-api/product-api-types.ts:
```diff
@@ -0,0 +1,167 @@
+
+type Prices = {
+  id: string;
+  value: { centAmount: number; currencyCode: string; fractionDigits: number; type: number };
```
add common type for value

pyropetrick commented on src/api/product-api/product-api.ts:
```diff
@@ -0,0 +1,73 @@
+import { CTP_PROJECT_KEY, CTP_API_URL, TOKEN_STORAGE_KEY } from '../api-data';
+import { AllProducts, FilterProductsQuery, Product } from './product-api-types';
+
+export class ProductAPI {
```
maybe try {} catch() {} on requests?)

GinezisNWD committed:
3.4.detailed product page (#34)

## PR Type

What kind of change does this PR introduce?

```
- [x] Feature
- [ ] Bugfix
- [ ] Refactoring
- [ ] Documentation changes
- [ ] Tests
- [ ] Other
```

## What's new?
Добавлена детальная страница товара, соответствует тех заданию всё,
кроме отсутствующего слайдера. Слайдер доделаю хотфиксом.

miroslav-zarenkov committed:
3.3.profile page (#35)

## PR Type

What kind of change does this PR introduce?

```
- [x] Feature
- [x] Bugfix
- [x] Refactoring
- [ ] Documentation changes
- [ ] Tests
- [ ] Other
```

## What's new?

Added profile page, refactored validation and register page.

toronto-tokyo committed:
3.2.4.catalog categorization (#36)

## PR Type

What kind of change does this PR introduce?

```
- [x] Feature
- [ ] Bugfix
- [ ] Refactoring
- [ ] Documentation changes
- [ ] Tests
- [ ] Other
```

## What's new?

Extended filters, add categorization to catalog

Margaryta-Maletz reviewed (approved):
Good job!
But I hope that in the next sprint you will do your best and finish earlier. And I am looking forward to unit tests!!!

Mentor Evaluation Criteria passed partially 45 / 60.
 - [x] Code Review Participation (10 points): Students actively participate in code reviews, providing constructive feedback and suggestions to their peers. 🗣️
 - [x] No Merges Without Approvals (10 points): Pull requests are not merged without approvals from other team members, ensuring the quality of code contributions. 👍
 - [x] Pull Request Descriptions (10 points): Pull requests include detailed descriptions, outlining the changes made and the purpose of the changes, facilitating better understanding and review. 📝
 - [x] Task Board Management (10 points): Students maintain and regularly update a task board (e.g., Trello, Jira), ensuring that tasks are moved according to their progress and completion. 📊
 - [x] Active Participation in Development Discussions (5 points): Students take an active role in discussing the development process, sharing ideas, and collaborating on solutions and improvements. 🗨️
 - [ ] Unit Test Coverage (15 points): The codebase has a minimum of 30% unit test coverage, ensuring the reliability and robustness of the implemented features. ✔️

GinezisNWD committed:
feat : add papaSlider (#38)

## PR Type

What kind of change does this PR introduce?

```
- [x] Feature
- [ ] Bugfix
- [ ] Refactoring
- [ ] Documentation changes
- [ ] Tests
- [ ] Other
```

## What's new?
Добавил папа слайдер, вроде даже адаптивный, вроде даже не багается.

toronto-tokyo committed:
3.2.4.catalog categorization fix (#37)

## PR Type

What kind of change does this PR introduce?

```
- [X] Feature
- [X] Bugfix
- [ ] Refactoring
- [ ] Documentation changes
- [ ] Tests
- [ ] Other
```

## What's new?

Implement reset button for filters. Add navigation into filters. Add
current filters view. Fix images url path in CatalogView. Add adaptive on catalog and category pages.

miroslav-zarenkov committed:
3.3.1.profile page fixes (#39)

## PR Type

What kind of change does this PR introduce?

```
- [x] Feature
- [x] Bugfix
- [x] Refactoring
- [ ] Documentation changes
- [ ] Tests
- [ ] Other
```

## What's new?

Change addresses handler in registration and profile.
